### DISCUSSION TITLE
Refining The Return Type Of Java Methods Without Breaking Backwards-Compatibility - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#286] [Announcements]
# Refining The Return Type Of Java Methods Without Breaking Backwards-Compatibility - Gunnar Morling

If you work on any kind of software library,
ensuring backwards-compatibility is a key concern:
if there’s one thing which users really dislike, it is breaking changes in a new version of a library.
The rules of what can (and cannot) be changed in a Java API without breaking existing consumers are well defined in the Java language specification (JLS),
but things can get pretty interesting in certain corner cases.

The Eclipse team provides a comprehensive overview about API evolution guidelines in their wiki.
When I shared the link to this great resource on Twitter the other day,
I received an interesting reply from Lukas Eder:

I wish Java had a few tools to prevent some cases of binary compatibility breakages. E.g. when refining a method return type, I’d like to keep the old method around in byte code (but not in source code).

I think kotlin has such tools?

In the remainder of this post,
I’d like to provide some more insight into that problem mentioned by Lukas,
and how it can be addressed using an open-source tool called Bridger.

https://www.morling.dev/blog/refining-return-type-java-methods-without-breaking-backwards-compatibility/

**Comment by gunnarmorling** [comment]
Learned about two other tools solving the same problem when discussing this post with folks on [Twitter](https://twitter.com/gunnarmorling/status/1462841605585518598) and [Reddit](https://www.reddit.com/r/java/comments/r15jjj/refining_the_return_type_of_java_methods_without/):

* [bridge-method-injector](https://github.com/infradna/bridge-method-injector) by Koshuke
* [OverloadReturn](https://github.com/JakeWharton/OverloadReturn) by Jake Wharton

**Comment by conzo77** [comment]
Thank you for writing an interesting article on a problem I was completely unaware of, and porividing not one but two alternative solutions.

It's moments like this when I enjoy being a Java dev even more than I usually do :)

**Reply by gunnarmorling** [reply]
Thanks a lot for the nice feedback 🙏 !

**Comment by mattwelke** [comment]
The subject confused me when I first encountered this on Reddit. I couldn't imagine a situation where one would need to keep the byte code the exact same. I forgot that Java supported loading classes at runtime, well after the application has started. Nice post.
